crypto/crypto/src/hash.rs

```diff
@@ -605,3 +616,101 @@ impl<T: ser::Serialize + ?Sized> TestOnlyHash for T {
         hasher.finish()
     }
 }
+
+/// Alternative to CryptoHash
+pub trait SimpleHash {
+    /// The hash method
+    fn hash(&self) -> HashValue;
+
+    #[cfg(any(test, feature = "fuzzing"))]
+    /// The salt used for hashing this type.
+    fn salt() -> Vec<u8>;
+}
+
+/// Cryptographic domains in Libra
+#[derive(Clone, Debug, PartialEq, Eq, Copy)]
+pub enum CryptoDomain {
+    /// Testing
+    Testing,
+    /// Common domain
+    Common,
+    /// Libra transactions
+    Transaction,
+    /// LibraBFT
+    Consensus,
+    /// Network protocols
+    Network,
+}
+
+/// A type whose hashing implementation is derived from LCS and a domain name.
+/// Because we use the Serde name of the type for seeding, only structs and enums are supported.
+pub trait CanonicalHash {
+    /// The crypto domain.
+    const DOMAIN: CryptoDomain;
+}
+
+/// A type that can store an initial hashing state to avoid recomputations.
+pub trait InitialState<State> {
+    /// Where to store the initial hashing state.
+    fn initial_state() -> &'static once_cell::sync::OnceCell<State>;
+}
+
+impl CryptoDomain {
+    /// The seed of the each domain value.
+    pub fn as_bytes(self) -> &'static [u8] {
+        use CryptoDomain::*;
+        match self {
+            Testing => b"Testing",
+            Common => b"Common",
+            Transaction => b"Transaction",
+            Consensus => b"Consensus",
+            Network => b"Network",
+        }
+    }
+}
+
+fn compute_canonical_salt<'de, T>() -> Vec<u8>
+where
+    T: CanonicalHash + serde::Serialize + serde::Deserialize<'de>,
+{
+    println!("Called for {}", std::any::type_name::<T>());
```

Comment:
Change this to a log line?

Reply:
oh my bad, I was supposed to remove this